41 spherical lensing

ASKabalan opened the pull request.

EiffL commented:
@ASKabalan just a few small comments on good practices:

- remember to add a description of what your PR is doing in the initial post. Here for instance, if you are adding some lensing functionality, you can show a plot or some sort of results of validation tests to give the reviewer a sense of how well the added feature is working.

- it is good practice to keep your PRs atomistic, around well defined scope. Here for instance I see modifications to growth.py, a simplectic integrator, and some lensing utilities. These are pretty unrelated things, it would be better to split into several PRs

EiffL commented on jaxpm/spherical.py:
```diff
@@ -0,0 +1,57 @@
+import healpy as hp
+import jax
+import jax.numpy as jnp
+import jax_healpy as jhp
```
Please add the dependency to the pyrpoject.toml

EiffL reviewed (changes requested):
- I would suggest changing the approach to bin directly the particles on the spherical maps

- Please remove from this PR things that are not related to spherical lensing, you can put them into a separate PR

- We need to add validation of the spherical lensing maps from somewhere, maybe comparing to glass?

EiffL commented:
Ignore last message ^^

ASKabalan commented:
@EiffL 

# Branch History
> remember to add a description of what your PR is doing in the initial post. Here for instance, if you are adding some lensing functionality, you can show a plot or some sort of results of validation tests to give the reviewer a sense of how well the added feature is working.
#
> it is good practice to keep your PRs atomistic, around well defined scope. Here for instance I see modifications to growth.py, a simplectic integrator, and some lensing utilities. These are pretty unrelated things, it would be better to split into several PRs


You are right .. the other functionalities come from other branch since I created this one out of the fix-sharding branch
I can rewrite history and make sure that  it matches that of main .. but we all know what do you think of that :smiling_face_with_tear: 

>I would suggest instead to bin the particles directly onto the spherical maps? that would be a much better approach.

Yeah I agree ..  I think a find a way to do it better in the litterature

> we can see a very weird pattern on these maps that probably has something to do iwth the 3d box

Yes I figured out that my  D_R goes outwards which means that the first projection includes only the corner of the box since it is longer that the box depth

https://github.com/DifferentiableUniverseInitiative/JaxPM/blob/09d82283f5600de3ddc0070081ebd66f2e5a3c48/jaxpm/spherical.py#L33

This explains that wierd circles .. I just need to integrate the D_R backwards

>Please remove from this PR things that are not related to spherical lensing, you can put them into a separate PR

Can I "slightly" rewrite history of this branch then :pray:

EiffL commented:
let's discuss the caching in a different thread, I can tell you what my logic was.

EiffL commented:
I'll stop playing with this for now ^^ I let you integrate it with lensing and adding validation tests against glass to make sure we are getting consistent Born raytracing with glass at least

ASKabalan commented:
I think this works now 

![image](https://github.com/user-attachments/assets/61025ae8-48fa-44a5-8102-2edce69dd94f)

check notebook https://github.com/DifferentiableUniverseInitiative/JaxPM/blob/41-spherical-lensing/notebooks/07-SphericalRayTracing.ipynb

EiffL commented:
So, how is it going :-) ?

ASKabalan commented:
I am almost there 
Was struggeling with units of jc.background.radial_comoving_distance
it seems to be Mpc/h
But sometimes people divide by cosmo.h to get it in Mpc
for example https://github.com/LSSTDESC/glass-jax/blob/914451146adaba590eacc23422ce443ebe72a5a3/glass/_src/shells.py#L100

I checked with this https://www.astro.ucla.edu/%7Ewright/CosmoCalc.html
`jc.background.radial_comoving_distance(cosmo, a) / cosmo.h` is in Mpc

I will add a plot of CL in the conversation

EiffL commented:
This is great progress @ASKabalan ! 

Remember to work in incremental atomic PRs, we can move this discussion to the lensing issue, because here you are testing the impact of different resolutions, but that's a different question from testing if your lensing kernel is working.


I would recommend making a first PR where you implement the spherical binning and lensing operation, and include a comparison against glass as a unit test. This test would be independent from the resolution of the simulation, the question it answers is whether you get the right convergence for a given set of density maps.

EiffL commented:
And that gradients work... Come to think of it, the nearest cell binning we are doing may not be differentiable.... We might have to implement the equivalent of a cic kernel on the spherical mesh... This needs to be checked

ASKabalan commented on jaxpm/spherical.py:
```diff
@@ -0,0 +1,153 @@
+        sums = jnp.bincount(pixels, weights=voxels, length=npix)
+        return sums
+
+    sum_map = interpolate_volume(coords, volume, pixels).sum(axis=0)
```
to be removed ... it was the Density to Healpix painting which is a double binning as you said
I will remove it

EiffL commented:
> 1 - Would you say that the CL coming from the spherical lensing is good enough ?
No they are not good enough, there seems to be a systematically high bias at low ell. In terms of ell_max, we also need to go further (see Cyrile's paper)
![image](https://github.com/user-attachments/assets/a73246e0-9003-4486-bde4-e997e45af381)

> 2 - For flat sky let's not worry about it ever? because the cl seems very off
I don't know exactly what evaluation you ran but flat sky certainly is working, c.f. https://github.com/EiffL/LPTLensingComparison/blob/main/notebooks/LPTLensingAccuracyTest.ipynb
![image](https://github.com/user-attachments/assets/0ea8e6e6-9b6c-4e68-b4dd-cea2c4c49a7f)

ASKabalan commented:
I read cyril's paper and it seems to focus on the shear map mainly on modelisation and the E/B purification which not our problem (for now)

Concerning the overestimation of the power at low ellms .. I did some reading and it could be the fact that the limber approximations (which is used by jax cosmo) can fail at low ells for a very low redshift

I want to compare with pyccl with beyond limber approximation
Do you think this is a waste of time
Anyways with the CIC I did and a appropriate nside value I get

![image](https://github.com/user-attachments/assets/876b08b9-2e16-45a3-95b2-6a9a586d61e8)

Do you consider the missmatch at low ells to be very bad ?
Comparing with a non limber approx is a waste of time?
Should I move on and consider the lensing spectra correct ?

EiffL commented:
At ell of around 100 we want to have exactly (at the subpercent level) the right amplitude.


But again, it's a differnt question, it's a matter of resolution of the simulation and binning schemes, not a question of correctness of the lensing operation.

EiffL commented:
oh and yeah comparing with non limber is not a priortiy at all :-)

ASKabalan commented:
> hummmmmm why is 128 better than 256 ^^? quick note also, it might be a bit dangerous to look at just one realization of these PSs, see my notebook here where I make a few realizations and plot the mean and std: https://github.com/EiffL/LPTLensingComparison/blob/main/notebooks/LPTLensingAccuracyTest.ipynb
> 
> It's true that limber can deviate below ell of 50 or so. But we probably wont use these scales anyway. If you look at cryrille's plot, the first data point is around ell of 30 I think

I projected 128 to nside 64 and 256 to nside 128

Should I compare with same nside ?

EiffL commented:
well i dont know, what we want is that whatever scheme you choose, you should get the mean of a bunch of realisations to match the theory. in the plot above, looks like blue is better than orange in the range of scales where both are valid.

ASKabalan commented:
I am transforming this into a playground branch to try do the lensing end to end and discuss

Actual dev is split into 3 PR

- An appropriate spherical binning https://github.com/DifferentiableUniverseInitiative/JaxPM/issues/45
- Validating against born (independant of binning)  https://github.com/DifferentiableUniverseInitiative/JaxPM/issues/47
- Validating against theory (once both are done)

ASKabalan commented:
@EiffL 

I think this PR Is also good to go (I am doing final touches)

But in short maps seems to be good

<img width="1010" height="809" alt="image" src="https://github.com/user-attachments/assets/12d7c57f-2a6a-44e5-9e6e-72ebc9d83481" />


The differences are acceptable

<img width="1010" height="809" alt="image" src="https://github.com/user-attachments/assets/d5d980c4-56b6-40af-8dbe-dee8e78e84ed" />

And cl ratios between jaxpm and glass tophat is within 10%

<img width="1789" height="1145" alt="image" src="https://github.com/user-attachments/assets/3a7630b5-75ee-4597-820b-3f2679bd651c" />

I am adding a test

ASKabalan committed:
add lensing and basic Spherical painting functionality

ASKabalan committed:
Allow Cosmo Cache deactivation

ASKabalan committed:
add Symplectic ODE terms for fast PM integration

ASKabalan committed:
add test against glass convergence maps

ASKabalan committed:
format

ASKabalan committed:
add glass[examples] to requirements-test.txt

ASKabalan committed:
format

ASKabalan committed:
format

ASKabalan committed:
add Jax healpy as a dependency

ASKabalan committed:
Roll back growth.py cache

ASKabalan committed:
Update testing dependencies

ASKabalan committed:
update lensing.py and test_convergence_vs_glass to fix jax-cosmo cache issue

ASKabalan committed:
Revert the fastpm odes to use Growth functions

ASKabalan committed:
Add safe division by zero in lensing.py

ASKabalan committed:
add save division by zero in spherical.py

ASKabalan committed:
Allow deactivating the usage of growth factors in ode.py in the FastPM formula

ASKabalan committed:
Merge branch 'main' into 41-spherical-lensing

ASKabalan committed:
Fix flat sky lensing issues and update the painting method to use the latests tecnhiques.

ASKabalan committed:
merge remote-tracking branch 'origin/main' into 41-spherical-lensing

ASKabalan committed:
Merge remote-tracking branch 'origin/main' into 41-spherical-lensing

ASKabalan committed:
Add an analytical function to compute visibility mask from observer position

ASKabalan committed:
Merge branch 'main' into 41-spherical-lensing

ASKabalan committed:
format

ASKabalan committed:
Update jax.experimental.shard_map to jax.shard_map and bump jax version

ASKabalan committed:
format

ASKabalan committed:
Update shardmap signature

Copilot committed:
Update JAX dependencies and migrate from experimental shard_map API (#52)

* Update minimum jaxdecomp version to v0.2.9

Co-authored-by: ASKabalan <83787080+ASKabalan@users.noreply.github.com>

ASKabalan committed:
Merge remote-tracking branch 'origin/update-shardmap-import' into 41-spherical-lensing

ASKabalan committed:
Update test workflow install order

ASKabalan committed:
Merge remote-tracking branch 'origin/update-shardmap-import' into update-shardmap-import

ASKabalan committed:
Add Set type import in distributed.py

ASKabalan committed:
Merge remote-tracking branch 'origin/update-shardmap-import' into 41-spherical-lensing

ASKabalan committed:
add

ASKabalan committed:
temp commit

ASKabalan committed:
Commit

ASKabalan committed:
update jax-cosmo usage to match update in ASKabalan:add_growth_second_to_cache

ASKabalan committed:
Simplify spherical painting (flatten later to use more devices)

ASKabalan committed:
small fix

ASKabalan committed:
remove print statements used for debugging

ASKabalan committed:
update spherical and 2d painting to work better for sharding

ASKabalan committed:
Remove prints and add jnp.asarray in utils

ASKabalan committed:
Update painting to better work for sharding

ASKabalan committed:
Update spherical painting to be sharded

ASKabalan committed:
format

ASKabalan committed:
Allow 1d and 2d normal fields in distributed mode

ASKabalan committed:
Update docstrings for sharding in spherical.py

ASKabalan committed:
Add test for normal_field

ASKabalan committed:
temporary CI fix

ASKabalan committed:
temporary CI fix

ASKabalan committed:
temporary CI fix

ASKabalan committed:
temporary CI fix

ASKabalan committed:
Migrate kernels.py to new jaxdecomp API and fix .at[] bug

Migrate from get_output_specs to get_fft_output_sharding (new jaxdecomp).
Fix gradient_kernel(order=0) which mutated a JAX array in place.
Simplify interpolate_power_spectrum to drop a redundant reshape.
Also tighten get_halo_size to accept a per-axis tuple.

ASKabalan committed:
Refactor growth.py to match jax_cosmo PR #143

Replace local _growth_factor_ODE caches with jax_cosmo's
_compute_growth_tables. Drop the cosmo._workspace dict (no longer
exists upstream). Update tests that wrote cosmo._workspace = {}.
Pin CI to ASKabalan/jax_cosmo@fix-import-error until upstream
PR #146 merges.

ASKabalan committed:
Add symplectic FastPM ODE integrators

Add symplectic_fpm_ode (drift/kick/first_kick) and a simpler
symplectic_ode. Both can optionally use the growth functions
(Gf / dGfa / gp) instead of pure scale-factor ratios for the
drift and kick factors.

ASKabalan committed:
Rewrite lensing.py for sharded painting

Replace density_plane with density_plane_fn (closure-based RHS
for a diffeqsolve), add spherical_density_fn that uses the new
sharded paint_particles_spherical, and update CIC paint calls
for the shape-preserving cic_paint_2d.

Add 08-convergence-vs-glass notebook for validation against GLASS.

ASKabalan committed:
CI: switch jax_cosmo install to better-cache branch

The refactored growth.py uses _compute_growth_tables and the new
tuple-based cache, which live on ASKabalan:better-cache rather than
fix-import-error.

ASKabalan committed:
CI: switch jax_cosmo install to better-cache branch

ode.py imports gp from jaxpm.growth, which only exists in the
refactored growth.py — that refactor depends on the upstream
better-cache fixes.

ASKabalan committed:
CI: switch jax_cosmo install to better-cache branch

This branch carries a partial growth.py refactor that uses
_compute_growth_tables; pin CI to ASKabalan:better-cache so the
expected cache symbols are available.

ASKabalan committed:
CI: switch jax_cosmo install to better-cache branch

Inherits the partial growth refactor from better-sharded-spherical-painting,
so the same upstream cache fixes are required.

ASKabalan committed:
Merge extend-normal_field for AxisType.Auto test fix

ASKabalan committed:
Merge extend-normal_field for AxisType.Auto test fix

Resolved tests.yml in favor of @better-cache (this branch already
includes the growth.py rewrite that requires the new odeint signature).

ASKabalan committed:
Merge extend-normal_field for AxisType.Auto test fix

Resolved tests.yml in favor of @better-cache (this branch rewrites
growth.py for the new odeint signature).

ASKabalan committed:
Merge update-growth-to-match-jax-cosmo for new odeint signature

Resolved growth.py in favor of #55's full rewrite (which also includes
the gp() function added by this branch).

ASKabalan committed:
Merge update-growth-to-match-jax-cosmo for new odeint signature

Resolved growth.py in favor of #55's full rewrite.

ASKabalan committed:
Halo is now a tuple as requested by the API

ASKabalan committed:
Fix spherical painting test and pin jax_healpy to be from main branch
until a new release (0.6.1+) that includes
the multi-dim batch shape fix is out.

ASKabalan committed:
remove 3.10 from CI test matrix add 3.13

ASKabalan committed:
remove 3.10 from CI test matrix add 3.13

ASKabalan committed:
remove 3.10 from CI test matrix add 3.13

ASKabalan committed:
remove 3.10 from CI test matrix add 3.13

ASKabalan committed:
use growth is now mandatory for fpm odes

ASKabalan committed:
Merge branch 'jaxdecomp-fixes' into better-sharded-spherical-painting

ASKabalan committed:
refactor all odes into ode.py

ASKabalan committed:
Merge remote-tracking branch 'origin/better-sharded-spherical-painting' into 41-spherical-lensing

ASKabalan committed:
Merge branch 'fastpm-odes' into 41-spherical-lensing

ASKabalan committed:
Merge branch 'update-growth-to-match-jax-cosmo' into 41-spherical-lensing

ASKabalan committed:
Finish the notebook

ASKabalan committed:
Fix the analytical visibility mask for the spherical painting method (FINALLY WORKING).
